src/pykeen/triples/splitting.py

```diff
@@ -525,3 +542,116 @@ def split(
         ratios=ratios,
         random_state=random_state,
     )
+
+
+def _entity_mask(mapped_triples: torch.Tensor, heads: torch.Tensor, tails: torch.Tensor) -> torch.Tensor:
+    head_mask = torch.isin(mapped_triples[:, COLUMN_HEAD], test_elements=heads)
+    tail_mask = torch.isin(mapped_triples[:, COLUMN_TAIL], test_elements=tails)
+    return head_mask & tail_mask
+
+
+def split_semi_inductive(
+    mapped_triples: MappedTriples,
+    *,
+    ratios: float | Sequence[float] = 0.8,
+    random_state: TorchRandomHint = None,
+) -> Sequence[MappedTriples]:
+    """Create a semi-inductive split, as defined by [ali2021]_.
+
+    In a semi-inductive split, we first split the entities into training and evaluation entities.
+    The training graph is then composed of all triples involving only training entities.
+    The evaluation graphs are built by looking at the triples that involve exactly one training
+    and one evaluation entity.
+
+    :param mapped_triples: shape: (n, 3)
+        The ID-based triples.
+    :param ratios:
+        The *entity* split ratio(s).
+    :param random_state:
+        The random state used to shuffle and split the triples.
+
+    :return:
+        A partition of triples, which are split (approximately) according to the ratios, stored TriplesFactory's
+        which share everything else with this root triples factory.
+    """
+    # normalize input
+    generator = ensure_torch_random_state(random_state=random_state)
+    ratios = normalize_ratios(ratios=ratios)
+
+    # split entities
+    train, *other = _random_split_unique_values(mapped_triples[:, 0::2], ratios=ratios, generator=generator)
+    sizes = [len(entities) for entities in [train, *other]]
+    logger.info(f"Entity split into groups of {sizes=}")
+
+    # extract training triples
+    result = [mapped_triples[_entity_mask(mapped_triples=mapped_triples, heads=train, tails=train)]]
+
+    # extract other triples
+    for other_entities in other:
+        train_head_to_other_tail = _entity_mask(mapped_triples=mapped_triples, heads=train, tails=other_entities)
```

Comment:
is there a more efficient way to do both of these entity masks at the same time?

Reply:
I couldn't see one.

Reply:
on second look also I don't think there is. I was just dreaming since there's some nice symmetry